File: ios/GlassSwitchView.swift
import ExpoModulesCore
import UIKit

/// iOS 26 style switch. Pressing turns the white thumb into a clear lens over the track, dragging
/// moves it and blends the track colour, releasing settles it on the nearest side.
final class GlassSwitchView: ExpoView, UIGestureRecognizerDelegate {
  let onValueChange = EventDispatcher()

  var value = false
  var disabled = false
  var onColor: UIColor?
  var lens = true

  private let track = CALayer()
  private let trackMask = CAShapeLayer()
  private let thumb = CALayer()
  private let lensView = UIView()
  private let lensRim = CAGradientLayer()
  private let lensRimMask = CAShapeLayer()
  private var shader: LensShaderRenderer?
  // plain magnified copy when Metal is not available
  private let fallbackTrack = CALayer()

  // thumb position 0..1, press 0..1
  private lazy var springs = SpringDriver(count: 2) { [weak self] in self?.applyFrame() }
  private var placed = false
  private var dragging = false
  private var moved = false
  private var downX: CGFloat = 0
  private var downPos: CGFloat = 0

  private lazy var touch: UILongPressGestureRecognizer = {
    let r = UILongPressGestureRecognizer(target: self, action: #selector(handleTouch(_:)))
    r.minimumPressDuration = 0
    r.allowableMovement = .greatestFiniteMagnitude
    r.delegate = self
    return r
  }()

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = false
    trackMask.fillRule = .evenOdd
    thumb.backgroundColor = UIColor.white.cgColor
    thumb.shadowColor = UIColor.black.cgColor
    thumb.shadowOpacity = 0.15
    thumb.shadowRadius = 3
    thumb.shadowOffset = CGSize(width: 0, height: 1.5)
    if #available(iOS 13.0, *) {
      track.cornerCurve = .continuous
      thumb.cornerCurve = .continuous
      lensView.layer.cornerCurve = .continuous
    }
    lensView.isUserInteractionEnabled = false
    lensView.clipsToBounds = true
    lensView.isHidden = true
    // glass edges show as light, glints top-left and bottom-right only
    lensRim.type = .conic
    lensRim.startPoint = CGPoint(x: 0.5, y: 0.5)
    lensRim.endPoint = CGPoint(x: 1, y: 0.5)
    lensRim.locations = [0, 0.125, 0.3, 0.45, 0.625, 0.8, 1]
    // transparent white, UIColor.clear is black and greys the fade
    let clear = UIColor(white: 1, alpha: 0).cgColor
    lensRim.colors = [
      clear, UIColor(white: 1, alpha: 0.7).cgColor, clear, clear, UIColor(white: 1, alpha: 0.95).cgColor, clear, clear,
    ]
    lensRimMask.fillColor = nil
    lensRimMask.strokeColor = UIColor.black.cgColor
    lensRimMask.lineWidth = 1.5
    lensRim.mask = lensRimMask
    lensView.layer.addSublayer(fallbackTrack)
    lensView.layer.addSublayer(lensRim)
    layer.addSublayer(track)
    addSubview(lensView)
    // drawn over the lens, so on release its fade-in hides the shrinking lens
    layer.addSublayer(thumb)
    addGestureRecognizer(touch)
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    applyFrame()
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    if window == nil {
      springs.stop()
      shader?.release()
    }
  }

  override func traitCollectionDidChange(_ previous: UITraitCollection?) {
    super.traitCollectionDidChange(previous)
    applyFrame()
  }

  func propsDidUpdate() {
    alpha = disabled ? 0.45 : 1
    touch.isEnabled = !disabled
    let target: CGFloat = value ? 1 : 0
    if !placed {
      springs.snap(0, target)
      springs.snap(1, 0)
      placed = true
      applyFrame()
    } else if !dragging {
      settle(target)
    }
  }

  private func settle(_ pos: CGFloat) {
    springs.target[0] = pos
    springs.target[1] = 0
    springs.stiffness = 420
    springs.damping = 1
    springs.start()
  }

  // MARK: - geometry

  private var inset: CGFloat { 2 }
  private var restSize: CGSize {
    let h = bounds.height - 2 * inset
    return CGSize(width: h * 1.55, height: h)
  }

  private func trackColor(_ pos: CGFloat) -> UIColor {
    let dark = traitCollection.userInterfaceStyle == .dark
    let off = dark ? UIColor(red: 57 / 255, green: 57 / 255, blue: 61 / 255, alpha: 1) : UIColor(white: 233 / 255, alpha: 1)
    let on = onColor ?? UIColor(red: 52 / 255, green: 199 / 255, blue: 89 / 255, alpha: 1)
    var (r1, g1, b1, a1): (CGFloat, CGFloat, CGFloat, CGFloat) = (0, 0, 0, 0)
    var (r2, g2, b2, a2): (CGFloat, CGFloat, CGFloat, CGFloat) = (0, 0, 0, 0)
    off.getRed(&r1, green: &g1, blue: &b1, alpha: &a1)
    on.getRed(&r2, green: &g2, blue: &b2, alpha: &a2)
    let t = min(max(pos, 0), 1)
    return UIColor(red: r1 + (r2 - r1) * t, green: g1 + (g2 - g1) * t, blue: b1 + (b2 - b1) * t, alpha: 1)
  }

  private func thumbRect(pos: CGFloat, press: CGFloat) -> CGRect {
    let rest = restSize
    let lensH = bounds.height * 1.4
    let lensW = lensH * 1.5
    let w = rest.width + (lensW - rest.width) * press
    let h = rest.height + (lensH - rest.height) * press
    let travel = bounds.width - 2 * inset - rest.width
    let cx = inset + rest.width / 2 + travel * pos
    return CGRect(x: cx - w / 2, y: bounds.midY - h / 2, width: w, height: h)
  }

  private func applyFrame() {
    guard bounds.width > 0, bounds.height > 0 else { return }
    let pos = springs.value[0]
    let press = lens ? springs.value[1] : 0
    let rect = thumbRect(pos: pos, press: press)
    let radius = rect.height / 2
    let color = trackColor(pos)

    CATransaction.begin()
    CATransaction.setDisableActions(true)

    track.frame = bounds
    track.cornerRadius = bounds.height / 2
    track.backgroundColor = color.cgColor

    thumb.frame = rect
    thumb.cornerRadius = radius
    thumb.opacity = Float(1 - press)
    thumb.shadowPath = UIBezierPath(roundedRect: CGRect(origin: .zero, size: rect.size), cornerRadius: radius).cgPath

    let lensUp = press > 0.001
    lensView.isHidden = !lensUp
    if lensUp {
      // masked under the lens so the track doesn't show twice
      let hole = UIBezierPath(rect: bounds)
      hole.append(UIBezierPath(roundedRect: rect, cornerRadius: radius))
      trackMask.frame = bounds
      trackMask.path = hole.cgPath
      if track.mask == nil { track.mask = trackMask }

      lensView.frame = rect
      lensView.layer.cornerRadius = radius
      lensRim.frame = lensView.bounds
      lensRim.opacity = Float(press)
      lensRimMask.path = UIBezierPath(roundedRect: lensView.bounds.insetBy(dx: 0.75, dy: 0.75), cornerRadius: radius).cgPath
      drawLens(rect: rect, press: press, color: color)
    } else if track.mask != nil {
      track.mask = nil
    }

    CATransaction.commit()
  }

  private func drawLens(rect: CGRect, press: CGFloat, color: UIColor) {
    let magnification = 1 + 0.05 * press
    if let shader = shader ?? LensShaderRenderer(frame: .zero) {
      if self.shader == nil {
        self.shader = shader
        lensView.insertSubview(shader.view, at: 0)
      }
      fallbackTrack.isHidden = true
      let scale = window?.screen.scale ?? UIScreen.main.scale
      let size = bounds.size
      // the track is a flat capsule, cheap to redraw whenever its colour moves
      shader.load(size: size, scale: scale) { ctx in
        ctx.setFillColor(color.cgColor)
        ctx.addPath(UIBezierPath(roundedRect: CGRect(origin: .zero, size: size), cornerRadius: size.height / 2).cgPath)
        ctx.fillPath()
      }
      shader.view.frame = lensView.bounds
      let drawable = CGSize(width: (rect.width * scale).rounded(), height: (rect.height * scale).rounded())
      if shader.view.drawableSize != drawable { shader.view.drawableSize = drawable }
      var u = shader.uniforms
      u.center = SIMD2(Float(rect.midX * scale), Float(rect.midY * scale))
      u.magnification = Float(magnification)
      u.radius = Float(rect.height / 2 * scale)
      u.bezel = Float(rect.height * 0.25 * scale)
      u.refraction = Float(3 * scale * press)
      u.chroma = 1
      u.press = Float(press)
      shader.uniforms = u
      shader.render()
    } else {
      fallbackTrack.isHidden = false
      fallbackTrack.backgroundColor = color.cgColor
      fallbackTrack.bounds = bounds
      fallbackTrack.cornerRadius = bounds.height / 2
      fallbackTrack.setAffineTransform(CGAffineTransform(scaleX: magnification, y: magnification))
      fallbackTrack.position = CGPoint(
        x: lensView.bounds.midX - (rect.midX - bounds.midX) * magnification,
        y: lensView.bounds.midY - (rect.midY - bounds.midY) * magnification)
    }
  }

  // MARK: - touch

  @objc private func handleTouch(_ r: UILongPressGestureRecognizer) {
    let x = r.location(in: self).x
    let travel = max(1, bounds.width - 2 * inset - restSize.width)
    switch r.state {
    case .began:
      dragging = true
      moved = false
      downX = x
      downPos = springs.value[0]
      springs.target[1] = UIAccessibility.isReduceMotionEnabled ? 0 : 1
      springs.stiffness = 700
      springs.damping = 1
      springs.start()
    case .changed:
      let dx = x - downX
      if abs(dx) > 4 { moved = true }
      if moved {
        springs.target[0] = min(max(downPos + dx / travel, 0), 1)
        springs.start()
      }
    case .ended:
      dragging = false
      // a tap flips, a drag lands on whichever side the thumb is closer to
      let next = moved ? springs.target[0] > 0.5 : !value
      settle(next ? 1 : 0)
      if next != value {
        value = next
        onValueChange(["value": next])
      }
    default:
      dragging = false
      settle(value ? 1 : 0)
    }
  }

  // a scroll view around the switch waits for it
  func gestureRecognizer(_ g: UIGestureRecognizer, shouldBeRequiredToFailBy other: UIGestureRecognizer) -> Bool {
    other.view is UIScrollView
  }
}

File: android/src/main/java/expo/modules/adaptiveglass/GlassSwitchView.kt
package expo.modules.adaptiveglass

import android.content.Context
import android.content.res.Configuration
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Outline
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.RenderNode
import android.graphics.SweepGradient
import android.os.Build
import android.view.MotionEvent
import android.view.ViewConfiguration
import expo.modules.adaptiveglass.performance.GlassPerformanceMonitor
import expo.modules.adaptiveglass.renderers.AcrylicRenderer
import expo.modules.adaptiveglass.renderers.ShaderGlassRenderer
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import kotlin.math.abs
import kotlin.math.max

// iOS 26 style switch. Pressing turns the white thumb into a clear lens over the track, dragging
// moves it and blends the track colour, releasing settles it on the nearest side.
class GlassSwitchView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  val onValueChange by EventDispatcher<Map<String, Any>>()

  var value = false
  var disabled = false
  var onColor: Int? = null
  var lens = true
  // resolved light / dark from JS, see isNight
  var scheme = "system"

  private val density = resources.displayMetrics.density
  // thumb position 0..1, press 0..1
  private val springs = SpringSet(this, 2) { invalidate() }
  private var placed = false
  private var dragging = false
  private var moved = false
  private var downX = 0f
  private var downPos = 0f
  private val touchSlop = ViewConfiguration.get(context).scaledTouchSlop

  private val trackRect = RectF()
  private val thumbRect = RectF()
  private val trackPaint = Paint(Paint.ANTI_ALIAS_FLAG)
  private val thumbPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.WHITE }
  private val shadowPaint = Paint(Paint.ANTI_ALIAS_FLAG)
  private val path = Path()
  private val outline = Outline()
  private var lensNode: Any? = null // RenderNode on API 29+
  private var shader: Any? = null // ShaderGlassRenderer on API 33+
  private val glintMatrix = Matrix()
  private val glintPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.STROKE
    shader = SweepGradient(
      0f, 0f,
      intArrayOf(
        Color.TRANSPARENT, AcrylicRenderer.withAlpha(Color.WHITE, 0.7f), Color.TRANSPARENT,
        Color.TRANSPARENT, AcrylicRenderer.withAlpha(Color.WHITE, 0.95f), Color.TRANSPARENT,
        Color.TRANSPARENT,
      ),
      floatArrayOf(0f, 0.125f, 0.3f, 0.45f, 0.625f, 0.8f, 1f),
    )
  }

  init {
    setWillNotDraw(false)
    clipChildren = false
    clipToPadding = false
  }

  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    setMeasuredDimension(MeasureSpec.getSize(widthMeasureSpec), MeasureSpec.getSize(heightMeasureSpec))
  }

  override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) = Unit

  override fun onDetachedFromWindow() {
    springs.stop()
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) (lensNode as? RenderNode)?.discardDisplayList()
    super.onDetachedFromWindow()
  }

  fun propsDidUpdate() {
    alpha = if (disabled) 0.45f else 1f
    val target = if (value) 1f else 0f
    if (!placed) {
      springs.snap(0, target)
      springs.snap(1, 0f)
      placed = true
    } else if (!dragging) {
      settle(target)
    }
    invalidate()
  }

  private fun settle(pos: Float) {
    springs.target[0] = pos
    springs.target[1] = 0f
    springs.stiffness = 420f
    springs.damping = 1f
    springs.start()
  }

  private fun isDark() = isNight(scheme)

  private fun trackColor(pos: Float): Int {
    val on = onColor ?: Color.rgb(52, 199, 89)
    val off = if (isDark()) Color.rgb(57, 57, 61) else Color.rgb(233, 233, 234)
    val t = pos.coerceIn(0f, 1f)
    fun mix(a: Int, b: Int) = (a + (b - a) * t).toInt()
    return Color.rgb(
      mix(Color.red(off), Color.red(on)),
      mix(Color.green(off), Color.green(on)),
      mix(Color.blue(off), Color.blue(on)),
    )
  }

  private fun computeThumb(pos: Float, press: Float) {
    val inset = 2f * density
    val restH = height - 2 * inset
    val restW = restH * 1.55f
    val lensH = height * 1.4f
    val lensW = lensH * 1.5f
    val w = restW + (lensW - restW) * press
    val h = restH + (lensH - restH) * press
    val travel = width - 2 * inset - restW
    val cx = inset + restW / 2 + travel * pos
    thumbRect.set(cx - w / 2, height / 2f - h / 2, cx + w / 2, height / 2f + h / 2)
  }

  private fun drawTrack(canvas: Canvas, pos: Float) {
    trackRect.set(0f, 0f, width.toFloat(), height.toFloat())
    trackPaint.color = trackColor(pos)
    val r = height / 2f
    canvas.drawRoundRect(trackRect, r, r, trackPaint)
  }

  override fun onDraw(canvas: Canvas) {
    if (width == 0 || height == 0) return
    val pos = springs.value[0]
    val press = if (lens) springs.value[1] else 0f
    computeThumb(pos, press)
    val r = thumbRect.height() / 2

    // masked under the lens so the track doesn't show twice
    if (press > 0.001f) {
      path.reset()
      path.addRoundRect(thumbRect, r, r, Path.Direction.CW)
      val save = canvas.save()
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
        canvas.clipOutPath(path)
      } else {
        @Suppress("DEPRECATION")
        canvas.clipPath(path, android.graphics.Region.Op.DIFFERENCE)
      }
      drawTrack(canvas, pos)
      canvas.restoreToCount(save)
    } else {
      drawTrack(canvas, pos)
    }

    val solid = 1f - press
    if (solid > 0.001f) {
      shadowPaint.color = AcrylicRenderer.withAlpha(Color.BLACK, 0.12f * solid)
      canvas.drawRoundRect(
        thumbRect.left, thumbRect.top + 1.5f * density, thumbRect.right, thumbRect.bottom + 1.5f * density,
        r, r, shadowPaint,
      )
    }
    if (press > 0.001f) drawLens(canvas, pos, press, r)
    // drawn over the lens, so on release its fade-in hides the shrinking lens
    if (solid > 0.001f) {
      thumbPaint.alpha = (255 * solid).toInt()
      canvas.drawRoundRect(thumbRect, r, r, thumbPaint)
    }

    if (press > 0.001f) {
      glintMatrix.setTranslate(thumbRect.centerX(), thumbRect.centerY())
      glintPaint.shader.setLocalMatrix(glintMatrix)
      glintPaint.strokeWidth = 1.5f * density
      glintPaint.alpha = (255 * press).toInt()
      val half = glintPaint.strokeWidth / 2
      canvas.drawRoundRect(
        thumbRect.left + half, thumbRect.top + half, thumbRect.right - half, thumbRect.bottom - half,
        r - half, r - half, glintPaint,
      )
    }
  }

  private fun drawLens(canvas: Canvas, pos: Float, press: Float, radius: Float) {
    val m = 1f + 0.05f * press
    val cx = thumbRect.centerX()
    val cy = thumbRect.centerY()
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && canvas.isHardwareAccelerated) {
      val node = (lensNode as? RenderNode) ?: RenderNode("GlassSwitchLens").also { lensNode = it }
      val left = thumbRect.left.toInt()
      val top = thumbRect.top.toInt()
      val w = thumbRect.width().toInt()
      val h = thumbRect.height().toInt()
      node.setPosition(left, top, left + w, top + h)
      outline.setRoundRect(0, 0, w, h, radius)
      node.setOutline(outline)
      node.setClipToOutline(true)
      val rc = node.beginRecording(w, h)
      try {
        rc.translate(-left.toFloat(), -top.toFloat())
        rc.scale(m, m, cx, cy)
        drawTrack(rc, pos)
      } finally {
        node.endRecording()
      }
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
        val s = (shader as? ShaderGlassRenderer) ?: ShaderGlassRenderer().also { shader = it }
        node.setRenderEffect(
          s.effect(
            null, w.toFloat(), h.toFloat(), radius, h * 0.25f, 3f * density * press, 1f,
            -0.25f * w, -0.6f * h, 0f, 1f, press,
          ),
        )
      }
      canvas.drawRenderNode(node)
    } else {
      val save = canvas.save()
      path.reset()
      path.addRoundRect(thumbRect, radius, radius, Path.Direction.CW)
      canvas.clipPath(path)
      canvas.scale(m, m, cx, cy)
      drawTrack(canvas, pos)
      canvas.restoreToCount(save)
    }
  }

  override fun onTouchEvent(ev: MotionEvent): Boolean {
    if (disabled) return false
    val inset = 2f * density
    val restW = (height - 2 * inset) * 1.55f
    val travel = max(1f, width - 2 * inset - restW)
    when (ev.actionMasked) {
      MotionEvent.ACTION_DOWN -> {
        parent?.requestDisallowInterceptTouchEvent(true)
        dragging = true
        moved = false
        downX = ev.x
        downPos = springs.value[0]
        springs.target[1] = if (GlassPerformanceMonitor.reduceMotion(context)) 0f else 1f
        springs.stiffness = 700f
        springs.damping = 1f
        springs.start()
      }
      MotionEvent.ACTION_MOVE -> {
        val dx = ev.x - downX
        if (abs(dx) > touchSlop) moved = true
        if (moved) {
          springs.target[0] = (downPos + dx / travel).coerceIn(0f, 1f)
          springs.start()
        }
      }
      MotionEvent.ACTION_UP -> {
        dragging = false
        // a tap flips, a drag lands on whichever side the thumb is closer to
        val next = if (moved) springs.target[0] > 0.5f else !value
        settle(if (next) 1f else 0f)
        if (next != value) {
          value = next
          onValueChange(mapOf("value" to next))
        }
      }
      MotionEvent.ACTION_CANCEL -> {
        dragging = false
        settle(if (value) 1f else 0f)
      }
    }
    return true
  }
}
